### PR TITLE
Regridded output now same dtype as input

### LINKED ISSUE #99: regrid output is always float64
In a current example [in docs](https://pangeo-xesmf.readthedocs.io/en/latest/notebooks/Rectilinear_grid.html#Regrid-between-rectilinear-grids)

In short:

```python
import numpy as np
import xarray as xr
import xesmf as xe

ds = xr.tutorial.open_dataset(
    "air_temperature"
)

dr = ds["air"]  # get a DataArray
dr.dtype
# dtype('float32')

dr["lat"].values, dr["lon"].values

# (array([75. , 72.5, 70. , 67.5, 65. , 62.5, 60. , 57.5, 55. , 52.5, 50. ,
#         47.5, 45. , 42.5, 40. , 37.5, 35. , 32.5, 30. , 27.5, 25. , 22.5,
#         20. , 17.5, 15. ], dtype=float32),
#  array([200. , 202.5, 205. , 207.5, 210. , 212.5, 215. , 217.5, 220. ,
#         222.5, 225. , 227.5, 230. , 232.5, 235. , 237.5, 240. , 242.5,
#         245. , 247.5, 250. , 252.5, 255. , 257.5, 260. , 262.5, 265. ,
#         267.5, 270. , 272.5, 275. , 277.5, 280. , 282.5, 285. , 287.5,
#         290. , 292.5, 295. , 297.5, 300. , 302.5, 305. , 307.5, 310. ,
#         312.5, 315. , 317.5, 320. , 322.5, 325. , 327.5, 330. ],
#        dtype=float32))

ds_out = xr.Dataset(
    {
        "lat": (["lat"], np.arange(16, 75, 1.0, dtype="float32")),
        "lon": (["lon"], np.arange(200, 330, 1.5, dtype="float32")),
    }
)

regridder = xe.Regridder(ds, ds_out, "bilinear")

dr_out = regridder(dr)
dr_out.dtype
# dtype('float64')
```

I need manually convert this output back to `float32` a la `dr_out.astype("float32")`, otherwise my files get big. Might be worth handling this in `xesmf`?

This is an issue brought up earlier in https://github.com/JiaweiZhuang/xESMF/issues/66

Thanks for your time!

### PR DESCRIPTION
**Pull request by brews**
Output regridded data is now the same type as input data. If input is `xarray.Dataset` with multiple unique dtypes, raises `ValueError`.

PR includes integration tests for new behavior under different input types/containers.

One pre-existing test needed to be updated because it output `int`s when the test was expecting precision from `float`.

Close #99.

### CODE REVIEW

**Comment by brews** [comment]
I think this is ready for review!

I'd especially appreciate input in how this is throwing a `ValueError` if a multi-dtype `xr.Dataset` is input.

Also, the change I needed to make to the pre-existing unit test makes me a bit nervous because this PR is clearly breaking from past behavior. Just another thing to note. Anyone else nervous about this?

**Comment by brews** [comment]
I think I could use maintainer input on this. `pre-commit` is giving me some trouble. It looks like it might be a `pre-commit`/GH actions configuration issue?

Specifically, I'm looking at 
```
blackdoc.................................................................Failed
- hook id: blackdoc
- exit code: 1

Traceback (most recent call last):
  File "/home/runner/.cache/pre-commit/repotmc3t5mu/py_env-python3/bin/blackdoc", line 5, in <module>
    from blackdoc.__main__ import main
  File "/home/runner/.cache/pre-commit/repotmc3t5mu/py_env-python3/lib/python3.9/site-packages/blackdoc/__main__.py", line 10, in <module>
    from .blackcompat import find_project_root, gen_python_files, read_pyproject_toml
  File "/home/runner/.cache/pre-commit/repotmc3t5mu/py_env-python3/lib/python3.9/site-packages/blackdoc/blackcompat.py", line 9, in <module>
    import toml
ModuleNotFoundError: No module named 'toml'
```

from https://github.com/pangeo-data/xESMF/pull/102/checks?check_run_id=3091031309.

This PR hasn't added or changed dependencies.

I fixed other apparent linting issues by deactivating `pre-commit`.

**Comment by raphaeldussin** [comment]
@brews please feel free to add `toml` package to the various environment yml files under `ci`
it looks like `black` does not depend on `toml` anymore but `blackdoc` does:

https://github.com/keewis/blackdoc/issues/100

this should take care of the linting test

**Review by raphaeldussin — CHANGES_REQUESTED** [review]
sorry, this is not necessary anymore, fixed in master

**Comment by brews** [comment]
> @brews please feel free to add `toml` package to the various environment yml files under `ci`
> it looks like `black` does not depend on `toml` anymore but `blackdoc` does:
> 
> [keewis/blackdoc#100](https://github.com/keewis/blackdoc/issues/100)
> 
> this should take care of the linting test

`toml` is in the environments as of `87285ce`. This looks good to go on my end, unless CI/CD says otherwise.

**Comment by raphaeldussin** [comment]
the handling in the dataset is finicky and decisions need to be made in terms of desired behavior.
the rest looks good.

**Review by huard — APPROVED** [review]

**Comment by huard** [comment]
@raphaeldussin Could you clarify what needs to be done to merge this ? What decisions do we need to make?

**Comment by raphaeldussin** [comment]
@brews I have made a PR to your branch here: https://github.com/brews/xESMF/pull/2
passing a list of different dtypes in the dataset case works fine (see test), please check and update your PR if you are satisfied with the modifications then we can move forward and merge your PR into the main code.

**Review by raphaeldussin — CHANGES_REQUESTED** [review]
I'd recommend updating PR with modifs from here https://github.com/brews/xESMF/pull/2

File: xesmf/frontend.py
Comment (brews) [reply xesmf/frontend.py]: I merged in your PR. I had an error when I tried something like this earlier but this clearly works not. Thanks so much. Marking as resolved.

**Comment by brews** [comment]
Thanks for the bump, @huard. I merged in the suggested corrections from @raphaeldussin. I agree with @raphaeldussin that accepting multiple dtypes from Datasets is ideal behavior. It appears to work in testing so 👍 .

Pending CI/CD checks, I think this is ready for re-review, @raphaeldussin.

**Review by raphaeldussin — APPROVED** [review]

**Comment by raphaeldussin** [comment]
LGTM

**Pull request merged**
